File: web/src/types/agent.ts
export interface Agent {
  id: string
  org_id: string
  name: string
  description?: string
  model_config_id?: string
  subagent_model_config_id?: string
  system_prompt?: string
  skill_ids: string[]
  skills?: Skill[]
  tool_ids: string[]
  all_builtin_tools?: boolean
  tools?: Tool[]
  mcp_server_ids: string[]
  mcp_servers: MCPServerOrg[]
  folder_id?: string
  max_turns?: number
  max_subagents?: number
  max_subagent_turns?: number
  model_config_params?: Record<string, unknown>
  created_by: string
  created_at: string
  updated_at: string
}

export interface MCPServerOrg {
  id: string
  org_id: string
  name: string
  type: 'stdio' | 'http'
  command: string
  args?: string[]
  created_by: string
  created_at: string
  updated_at: string
}

export interface ModelConfig {
  id: string
  org_id: string
  name: string
  provider: string
  base_url: string
  model: string
  default_params?: Record<string, unknown>
  context_window: number
  price_per_input_token: number
  price_per_output_token: number
  price_per_cache_read_token: number
  folder_id?: string
  created_by: string
  created_at: string
  updated_at: string
}

export interface Skill {
  id: string
  org_id: string
  name: string
  description?: string
  system_prompt?: string
  folder_id?: string
  created_by: string
  created_at: string
  updated_at: string
}

export type ToolType = 'builtin' | 'webhook' | 'sql_query'

export interface Tool {
  id: string
  org_id: string
  name: string
  description: string
  type: ToolType
  schema: Record<string, any>
  config: Record<string, any>
  require_confirmation?: boolean
  folder_id?: string
  created_by: string
  created_at: string
  updated_at: string
}

export interface AgentSession {
  id: string
  agent_id: string
  notebook_id: string
  user_id: string
  max_turns: number
  title: string | null
  ended_at?: string
  created_at: string
}

export interface AgentMessage {
  id: string
  session_id: string
  role: 'user' | 'assistant' | 'tool' | 'compaction'
  content?: string
  tool_call_id?: string
  tool_calls?: ToolCall[]
  tokens_input?: number
  tokens_output?: number
  tokens_direct?: number
  image_ids?: string[]
  created_at: string
}

export interface ToolCall {
  id: string
  name: string
  arguments: Record<string, unknown>
  result?: unknown
  error?: string
  duration_ms?: number
  tokens_direct?: number
}

export interface SubagentTask {
  id: string
  goal: string
  status: 'queued' | 'running' | 'completed' | 'failed'
  result?: unknown
}

export interface AgentTaskItem {
  id: string
  description: string
  status: 'pending' | 'in_progress' | 'done'
}

export interface TokenBreakdown {
  input: number
  output: number
  reasoning: number
  cache_read: number
  model_calls: number
  system_prompt: number
  skill_override: number
  history: number
  user_message: number
  tool_definitions: number
  tool_calls: number
  tool_results: number
  context_current?: number
  subagent_input?: number
  subagent_output?: number
  duration_ms?: number
}

export interface SessionUsage {
  input: number
  output: number
  reasoning: number
  cache_read: number
  model_calls: number
  subagent_input: number
  subagent_output: number
  context_tokens: number
  context_window: number
}

export type WSMessage =
  | { type: 'token'; data: string; seq?: number }
  | { type: 'reasoning'; data: string; seq?: number }
  | { type: 'tool_call'; tool: string; tool_call_id?: string; params: string; args: unknown; result: unknown; reasoning?: string; duration_ms?: number; seq?: number }
  | { type: 'tool_result'; tool: string; tool_call_id?: string; params: string; result: string; error?: string; duration_ms?: number; tokens_direct?: number; seq?: number }
  | { type: 'cell_created'; cell_id: string; position: number; seq?: number }
  | { type: 'cell_output'; cell_id: string; outputs: Array<{ type: string; data: unknown }>; seq?: number }
  | { type: 'cell_updated'; cell_id: string; seq?: number }
  | { type: 'subagent_progress'; tasks: SubagentTask[] }
  | { type: 'subagent_status'; task_id: string; status: string; goal?: string; result?: unknown; error?: string; duration_ms?: number; tokens_input?: number; tokens_output?: number; session_usage?: SessionUsage | null }
  | { type: 'subagent_message'; task_id: string; role: string; content: string; result?: string; tool_call_id?: string; tool_calls?: any; reasoning_content?: string; duration_ms?: number }
  | { type: 'done'; tokens?: TokenBreakdown; data?: { content?: string; reasoning?: string; tokens?: TokenBreakdown; session_usage?: SessionUsage | null }; seq?: number }
  | { type: 'error'; message: string }
  | { type: 'slash_result'; command: string; data: unknown }
  | { type: 'backpressure_warning'; dropped_tokens: number }
  | { type: 'reconnect_sync'; messages: AgentMessage[] | null; running?: boolean; server_seq?: number; session_usage?: SessionUsage | null }
  | { type: 'tasks_updated'; data: AgentTaskItem[] }
  | { type: 'tool_confirm_required'; tool_name: string; tool_args: string; current_source?: string }
  | { type: 'resync'; seq?: number }
  | { type: 'llm_retry'; attempt?: number; max_attempts?: number; error?: string; seq?: number }
  | { type: 'steering'; content: string; seq?: number }
  | { type: 'steering_accepted'; seq?: number }
  | { type: 'steering_busy'; content: string; seq?: number }
  | { type: 'question'; question: string; options?: Array<{ title: string; description?: string } | string>; allow_custom: boolean }
  | { type: 'token_update'; tokens: TokenBreakdown; session_usage?: SessionUsage | null }
  | { type: 'context_compacted'; summary: string; tokens?: TokenBreakdown }
  | { type: 'cancelled' }

File: internal/models/agent.go
package models

import (
	"time"
)

type ModelConfig struct {
	ID                     string    `json:"id"`
	OrgID                  string    `json:"org_id"`
	Name                   string    `json:"name"`
	Provider               string    `json:"provider"`
	BaseURL                string    `json:"base_url"`
	Model                  string    `json:"model"`
	APIKeyEncrypted        []byte    `json:"-"`
	APIKeyEnvVar           *string   `json:"-"`
	DefaultParams          JSONMap   `json:"default_params"`
	ContextWindow          int       `json:"context_window"`
	PricePerInputToken     float64   `json:"price_per_input_token"`
	PricePerOutputToken    float64   `json:"price_per_output_token"`
	PricePerCacheReadToken float64   `json:"price_per_cache_read_token"`
	FolderID               *string   `json:"folder_id,omitempty"`
	CreatedBy              string    `json:"created_by"`
	CreatedAt              time.Time `json:"created_at"`
	UpdatedAt              time.Time `json:"updated_at"`
}

type Skill struct {
	ID           string    `json:"id"`
	OrgID        string    `json:"org_id"`
	Name         string    `json:"name"`
	Description  string    `json:"description,omitempty"`
	SystemPrompt string    `json:"system_prompt,omitempty"`
	FolderID     *string   `json:"folder_id,omitempty"`
	CreatedBy    string    `json:"created_by"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

type MCPServerOrg struct {
	ID        string    `json:"id"`
	OrgID     string    `json:"org_id"`
	Name      string    `json:"name"`
	Type      string    `json:"type"`
	Command   string    `json:"command"`
	Args      []string  `json:"args,omitempty"`
	CreatedBy string    `json:"created_by"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type ToolType string

const (
	ToolTypeBuiltin  ToolType = "builtin"
	ToolTypeWebhook  ToolType = "webhook"
	ToolTypeSQLQuery ToolType = "sql_query"
)

type Tool struct {
	ID                  string    `json:"id"`
	OrgID               string    `json:"org_id"`
	Name                string    `json:"name"`
	Description         string    `json:"description"`
	Type                ToolType  `json:"type"`
	Schema              JSONMap   `json:"schema"`
	Config              JSONMap   `json:"config"`
	RequireConfirmation bool      `json:"require_confirmation"`
	FolderID            *string   `json:"folder_id,omitempty"`
	CreatedBy           string    `json:"created_by"`
	CreatedAt           time.Time `json:"created_at"`
	UpdatedAt           time.Time `json:"updated_at"`
}

type Agent struct {
	ID                    string         `json:"id"`
	OrgID                 string         `json:"org_id"`
	Name                  string         `json:"name"`
	Description           string         `json:"description,omitempty"`
	ModelConfigID         *string        `json:"model_config_id,omitempty"`
	SubagentModelConfigID *string        `json:"subagent_model_config_id,omitempty"`
	SystemPrompt          string         `json:"system_prompt,omitempty"`
	SkillIDs              []string       `json:"skill_ids"`
	Skills                []Skill        `json:"skills,omitempty"`
	MCPServerIDs          []string       `json:"mcp_server_ids"`
	MCPServers            []MCPServerOrg `json:"mcp_servers"`
	ToolIDs               []string       `json:"tool_ids,omitempty"`
	AllBuiltinTools       bool           `json:"all_builtin_tools"`
	Tools                 []Tool         `json:"tools,omitempty"`
	ModelConfigParams     JSONMap        `json:"model_config_params,omitempty"`
	FolderID              *string        `json:"folder_id,omitempty"`
	MaxTurns              *int           `json:"max_turns,omitempty"`
	MaxSubAgents          int            `json:"max_subagents,omitempty"`
	MaxSubagentTurns      int            `json:"max_subagent_turns,omitempty"`
	CreatedBy             string         `json:"created_by"`
	CreatedAt             time.Time      `json:"created_at"`
	UpdatedAt             time.Time      `json:"updated_at"`
}

type AgentSession struct {
	ID         string     `json:"id"`
	AgentID    string     `json:"agent_id"`
	NotebookID string     `json:"notebook_id"`
	UserID     string     `json:"user_id"`
	MaxTurns   int        `json:"max_turns"`
	Title      *string    `json:"title,omitempty"`
	EndedAt    *time.Time `json:"ended_at,omitempty"`
	CreatedAt  time.Time  `json:"created_at"`
}

type ToolCall struct {
	ID         string  `json:"id"`
	Name       string  `json:"name"`
	Arguments  any     `json:"arguments"`
	Result     any     `json:"result,omitempty"`
	Error      *string `json:"error,omitempty"`
	DurationMs int     `json:"duration_ms,omitempty"`
}

type AgentMessage struct {
	ID               string     `json:"id"`
	SessionID        string     `json:"session_id"`
	Role             string     `json:"role"`
	Content          string     `json:"content,omitempty"`
	ToolCallID       *string    `json:"tool_call_id,omitempty"`
	ToolCalls        []ToolCall `json:"tool_calls,omitempty"`
	ReasoningContent string     `json:"reasoning_content,omitempty"`
	TokensInput      int        `json:"tokens_input,omitempty"`
	TokensOutput     int        `json:"tokens_output,omitempty"`
	TokensReasoning  int        `json:"tokens_reasoning,omitempty"`
	ModelCalls       int        `json:"model_calls,omitempty"`
	DurationMs       int        `json:"duration_ms,omitempty"`
	ImageIDs         []string   `json:"image_ids,omitempty"`
	CreatedAt        time.Time  `json:"created_at"`
}

type SubagentTask struct {
	ID              string     `json:"id"`
	ParentSessionID string     `json:"parent_session_id"`
	ParentMessageID *string    `json:"parent_message_id,omitempty"`
	AgentID         *string    `json:"agent_id,omitempty"`
	Goal            string     `json:"goal"`
	Context         JSONMap    `json:"context,omitempty"`
	Status          string     `json:"status"`
	Result          JSONMap    `json:"result,omitempty"`
	TokensInput     int        `json:"tokens_input,omitempty"`
	TokensOutput    int        `json:"tokens_output,omitempty"`
	CreatedAt       time.Time  `json:"created_at"`
	CompletedAt     *time.Time `json:"completed_at,omitempty"`
}

type AgentStatsDaily struct {
	Date          string `json:"date"`
	AgentID       string `json:"agent_id"`
	UserID        string `json:"user_id"`
	SessionsCount int    `json:"sessions_count"`
	MessagesCount int    `json:"messages_count"`
	TokensInput   int64  `json:"tokens_input"`
	TokensOutput  int64  `json:"tokens_output"`
}

type AgentVersion struct {
	ID            string    `json:"id"`
	AgentID       string    `json:"agent_id"`
	Version       int       `json:"version"`
	Name          *string   `json:"name,omitempty"`
	Description   *string   `json:"description,omitempty"`
	SystemPrompt  *string   `json:"system_prompt,omitempty"`
	SkillIDs      []string  `json:"skill_ids,omitempty"`
	ModelConfigID *string   `json:"model_config_id,omitempty"`
	ChangedBy     string    `json:"changed_by"`
	ChangeReason  string    `json:"change_reason,omitempty"`
	CreatedAt     time.Time `json:"created_at"`
}

type JSONMap map[string]any
